internal/logging/ring_buffer.go

```diff
@@ -0,0 +1,52 @@
+package logging
+
+// Ring buffer structure, but without a start position for reading. All reads are tail reads with
+// no concept for what has already been read. The use case here is for logging to keep track of the
+// last X bytes written.
+type ringBuffer struct {
+	buffer []byte
+	size   int
+	end    int
+	full   bool
```

Comment:
Can't all of these be extrapolated with `len(buffer)`?

Reply:
No, `len(buffer)` will always return the allocated size, not the number of bytes written to it. `end` revolves around the buffer as writes happen, and `full` indicates whether or not `end` has wrapped once.